onnxruntime/server/core/metric_registry.h

```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#include <prometheus/registry.h>
+
+namespace onnxruntime {
+namespace server {
+class MetricRegistry {
+public:
+  // returns the global metric registry instance
+  // use this handle to get the instance to increment/count requests
+  static MetricRegistry& Get() {
+      static MetricRegistry metricRegistry;
+      return metricRegistry;
+  }
+
+  // Store quantile buckets 
+  static prometheus::Histogram::BucketBoundaries TimeBuckets () {
+    // Anything greater than 10s we should bucket into +inf
+    return {0, 50, 100, 250, 500, 1000, 2500, 5000, 10000};
+  }
+
+  static prometheus::Histogram::BucketBoundaries ByteBuckets() {
+    // get megabyte boundaries?
+    return {
+      // 0, 0.5, 1, 2, 5, 10, 20 MiB
+      0, 524288, 1048576, 2097152, 5242880, 10485760, 20971520
+    };
+  }
+
+  // The internal prometheus registry
+  prometheus::Registry registry;
+  // The three metrics we're currently recording
+  // Inference Time (model and version can be handled by labels so you can)
+  // aggregate total onnxruntime server performance if desired
+  std::unique_ptr<prometheus::Family<prometheus::Histogram>*> inferenceTimer;
+  // Run Time (model and version again handled by labels)
+  std::unique_ptr<prometheus::Family<prometheus::Histogram>*> runTimer;
+  // Total number of HTTP requests split by path, this includes pinging the metric
+  // endpoint and the health checker
+  std::unique_ptr<prometheus::Family<prometheus::Counter>*> totalHTTPRequests;
+  // Total number of gRPC requests received by the server
+  std::unique_ptr<prometheus::Family<prometheus::Counter>*> totalGRPCRequests;
```

Comment:
Can `std::unique_ptr<prometheus::Family<prometheus::Counter>>` work?

Reply:
I've experimented with this but the mutex inside Prometheus Family is still causing issues.

Reply:
May I know more about the failure?

Reply:
```
/usr/local/opt/llvm/bin/../include/c++/v1/memory:3132:32: error: call to implicitly-deleted copy constructor of
      'prometheus::Family<prometheus::Histogram>'
    return unique_ptr<_Tp>(new _Tp(_VSTD::forward<_Args>(__args)...));
                               ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
$DEVELOPMENT_PATH/onnxruntime/server/core/metric_registry.h:61:27: note: in instantiation of function
      template specialization 'std::__1::make_unique<prometheus::Family<prometheus::Histogram>,
      prometheus::Family<prometheus::Histogram> &>' requested here
    inferenceTimer = std::make_unique<prometheus::Family<prometheus::Histogram>>(
                          ^
$DEVELOPMENT_PATH/onnxruntime/cmake/external/prometheus-cpp/core/include/prometheus/family.h:133:14: note: copy
      constructor of 'Family<prometheus::Histogram>' is implicitly deleted because field 'mutex_' has an inaccessible copy
      constructor
  std::mutex mutex_;
```

Reply:
Then let's just go with the raw pointer then. As long as MetricRegistry's lifetime is for the duration of the program it should be fine.